types/src/v2/error.rs

```diff
@@ -32,3 +58,147 @@ pub const INVALID_PARAMS_MSG: &str = "Invalid params";
 pub const INVALID_REQUEST_MSG: &str = "Invalid request";
 /// Method not found error message.
 pub const METHOD_NOT_FOUND_MSG: &str = "Method not found";
+/// Reserved for implementation-defined server-errors.
+pub const SERVER_ERROR_MSG: &str = "Server error";
+/// Application defined error which is not in the reserved space (-32000..=-32768)
+pub const APPLICATION_ERROR_MSG: &str = "Application error";
+
+/// JSONRPC error code
+#[derive(Error, Debug, PartialEq, Copy, Clone)]
+pub enum ErrorCode {
+	/// Invalid JSON was received by the server.
+	/// An error occurred on the server while parsing the JSON text.
+	ParseError,
+	/// The JSON sent is not a valid Request object.
+	InvalidRequest,
+	/// The method does not exist / is not available.
+	MethodNotFound,
+	/// Invalid method parameter(s).
+	InvalidParams,
+	/// Internal JSON-RPC error.
+	InternalError,
+	/// Reserved for implementation-defined server-errors.
+	ServerError(i32),
+	/// Application defined error which is not in the reserved space (-32000..=-32768)
+	ApplicationError(i32),
+}
+
+impl ErrorCode {
+	/// Returns integer code value
+	pub const fn code(&self) -> i32 {
+		match *self {
+			ErrorCode::ParseError => PARSE_ERROR_CODE,
+			ErrorCode::InvalidRequest => INVALID_REQUEST_CODE,
+			ErrorCode::MethodNotFound => METHOD_NOT_FOUND_CODE,
+			ErrorCode::InvalidParams => INVALID_PARAMS_CODE,
+			ErrorCode::InternalError => INTERNAL_ERROR_CODE,
+			ErrorCode::ServerError(code) => code,
+			ErrorCode::ApplicationError(code) => code,
+		}
+	}
+
+	/// Returns the message for the given error code.
+	pub const fn message(&self) -> &str {
+		match self {
+			ErrorCode::ParseError => PARSE_ERROR_MSG,
+			ErrorCode::InvalidRequest => INVALID_REQUEST_MSG,
+			ErrorCode::MethodNotFound => METHOD_NOT_FOUND_MSG,
+			ErrorCode::InvalidParams => INVALID_PARAMS_MSG,
+			ErrorCode::InternalError => INTERNAL_ERROR_MSG,
+			ErrorCode::ServerError(_) => SERVER_ERROR_MSG,
+			ErrorCode::ApplicationError(_) => APPLICATION_ERROR_MSG,
+		}
+	}
+}
+
+impl fmt::Display for ErrorCode {
+	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+		write!(f, "{}: {}", self.code(), self.message())
+	}
+}
+
+impl From<i32> for ErrorCode {
+	fn from(code: i32) -> Self {
+		match code {
+			PARSE_ERROR_CODE => ErrorCode::ParseError,
+			INVALID_REQUEST_CODE => ErrorCode::InvalidRequest,
+			METHOD_NOT_FOUND_CODE => ErrorCode::MethodNotFound,
+			INVALID_PARAMS_CODE => ErrorCode::InvalidParams,
+			INTERNAL_ERROR_CODE => ErrorCode::InternalError,
+			SERVER_ERROR_CODE_RANGE_START..=SERVER_ERROR_CODE_RANGE_END => ErrorCode::ServerError(code),
+			code => ErrorCode::ApplicationError(code),
+		}
+	}
+}
+
+impl<'a> serde::Deserialize<'a> for ErrorCode {
+	fn deserialize<D>(deserializer: D) -> Result<ErrorCode, D::Error>
```

Comment:
these are ser/deser is a little bit weird (not symmetrical) but essentially the error message can be deduced from the error code so no need to store the message as allocated string.